File: adventofcode.tests.csharp/2020/2020_15.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Xunit;

namespace Year2020
{
    public static class AOC2020_15
    {
        public static long SolveCase(string[] lines, int target)
        {
            Dictionary<int, List<int>> lastSpokenLookup = new();

            int[] startingNumbers = lines[0].Split(',').Select(int.Parse).ToArray();
            for (int i = 0; i < startingNumbers.Length; i++)
            {
                lastSpokenLookup[startingNumbers[i]] = new List<int> { i + 1 };
            }

            int curr = startingNumbers.Last();

            for (int i = startingNumbers.Length; i < target; i++)
            {
                int prev = curr;
                List<int> listForPrev = GetListForValue(lastSpokenLookup, prev);

                curr = listForPrev.Count <= 1
                    ? 0
                    : listForPrev[listForPrev.Count - 1] - listForPrev[listForPrev.Count - 2];

                List<int> listForCurr = GetListForValue(lastSpokenLookup, curr);
                listForCurr.Add(i + 1);
            }

            return curr;
        }

        static List<int> GetListForValue(Dictionary<int, List<int>> lookup, int value)
        {
            List<int> list;
            if (!lookup.TryGetValue(value, out list))
            {
                list = new List<int>();
                lookup.Add(value, list);
            }

            return list;
        }

        public static Tuple<long, long> Solve(string[] lines)
        {
            long ans1 = SolveCase(lines, 2020);
            long ans2 = SolveCase(lines, 30000000);

            return Tuple.Create(ans1, ans2);
        }
    }

    public class Day15
    {
        [Fact]
        public void Sample()
        {
            var lines = Utils.ReadInput("2020", "15", "sample");
            Utils.SolveAndValidate(Tuple.Create(436L, 175594L), AOC2020_15.Solve, lines);
        }

        [Fact]
        public void Actual()
        {
            var lines = Utils.ReadInput("2020", "15", "actual");
            Utils.SolveAndValidate(Tuple.Create(249L, 41687L), AOC2020_15.Solve, lines);
        }
    }
}

File: adventofcode.tests.fsharp/2020/2020_15.fs
namespace Year2020

open System.Collections.Generic

module Day15 =
    open Utils
    open Xunit

    let getList (lookup : Dictionary<int, int ResizeArray>) key =
        let found, lst = lookup.TryGetValue key
        if found then
            lst
        else
            let lst = ResizeArray([])
            lookup.[key] <- lst
            lst

    let solveCase target startingNums =
        let len = startingNums |> Array.length
        let lastSpokenLookup = new Dictionary<int, int ResizeArray>()

        [0..len-1] |> List.iter (fun i -> lastSpokenLookup.[startingNums.[i]] <- ResizeArray([i+1]))

        let mutable curr = startingNums |> Array.last

        for i in len..target-1 do
            let prev = curr
            let prevLst = getList lastSpokenLookup prev

            curr <-
                match prevLst.Count with
                | length when length <= 1 -> 0
                | length -> prevLst.[length-1] - prevLst.[length-2]

            let currLst = getList lastSpokenLookup curr
            currLst.Add(i+1)

        curr |> int64

    let solve (lines : string[]) =
        let starting = lines.[0].Split(',') |> Array.map int

        let ans1 = starting |> solveCase 2020
        let ans2 = starting |> solveCase 30000000

        (ans1, ans2)

    [<Fact>]
    let Sample () =
        readInput "2020" "15" "sample" |> solveAndValidate (436L, 175594L) solve

    [<Fact>]
    let Actual () =
        readInput "2020" "15" "actual" |> solveAndValidate (249L, 41687L) solve